File: app/src/main/java/com/stevenfrew/beatprompter/pref/ImageListPreference.java
package com.stevenfrew.beatprompter.pref;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.preference.ListPreference;
import android.preference.PreferenceManager;
import android.util.AttributeSet;
import android.app.AlertDialog.Builder;
import android.view.View;
import android.widget.ImageView;
import android.widget.ListAdapter;

import com.stevenfrew.beatprompter.ui.ImageArrayAdapter;
import com.stevenfrew.beatprompter.R;
import com.stevenfrew.beatprompter.SongList;

public class ImageListPreference extends ListPreference {
    private int[] resourceIds = null;

    public ImageListPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray typedArray = context.obtainStyledAttributes(attrs,
                R.styleable.ImageListPreference);

        int indexCount=typedArray.getIndexCount();
        Resources resources=context.getResources();
        String packageName=context.getPackageName();
        if(indexCount>0) {
            String[] imageNames = resources.getStringArray(
                    typedArray.getResourceId(indexCount - 1, -1));

            resourceIds = new int[imageNames.length];

            for (int i = 0; i < imageNames.length; i++) {
                String imageName = imageNames[i].substring(
                        imageNames[i].lastIndexOf('/') + 1,
                        imageNames[i].lastIndexOf('.'));

                resourceIds[i] = resources.getIdentifier(imageName,
                        "drawable", packageName);
            }

            typedArray.recycle();
        }
    }

    @Override
    public void onBindView(View view) {
        super.onBindView(view);
        ImageView imageView = view.findViewById(R.id.iconImageView);
        SharedPreferences sharedPrefs=PreferenceManager.getDefaultSharedPreferences(SongList.getContext());
        String value=sharedPrefs.getString(getKey(),"");

        if(value.equals(SongList.getContext().getString(R.string.googleDriveValue)))
            imageView.setImageResource(R.drawable.ic_google_drive);
        else if(value.equals(SongList.getContext().getString(R.string.dropboxValue)))
            imageView.setImageResource(R.drawable.ic_dropbox);
        else if(value.equals(SongList.getContext().getString(R.string.oneDriveValue)))
            imageView.setImageResource(R.drawable.ic_onedrive);
        else
            imageView.setImageResource(R.drawable.blank_icon);
    }

    private boolean inForceUpdate=false;
    public void forceUpdate()
    {
        if(!inForceUpdate) {
            try {

                inForceUpdate = true;
                notifyChanged();
            }
            finally
            {
                inForceUpdate=false;
            }
        }
    }

    /**
     * {@inheritDoc}
     */
    protected void onPrepareDialogBuilder(Builder builder) {
        int index = findIndexOfValue(getSharedPreferences().getString(
                getKey(), "1"));

        ListAdapter listAdapter = new ImageArrayAdapter(getContext(),
                R.layout.imagelistitem, getEntries(), resourceIds, index);

        // Order matters.
        builder.setAdapter(listAdapter, this);
        super.onPrepareDialogBuilder(builder);
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/ui/pref/ImageListPreference.kt
package com.stevenfrew.beatprompter.ui.pref

import android.content.Context
import android.util.AttributeSet
import android.widget.ImageView
import androidx.preference.ListPreference
import androidx.preference.PreferenceViewHolder
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.stevenfrew.beatprompter.BeatPrompter
import com.stevenfrew.beatprompter.Preferences
import com.stevenfrew.beatprompter.R

class ImageListPreference(private val mContext: Context, attrs: AttributeSet) :
	ListPreference(mContext, attrs) {
	private lateinit var resourceIds: IntArray

	private var inForceUpdate = false

	init {
		val typedArray = context.obtainStyledAttributes(
			attrs,
			R.styleable.ImageListPreference
		)

		entries = getFilteredCloudStorageEntries()
		entryValues = getFilteredCloudStorageEntryValues()

		val indexCount = typedArray.indexCount
		val resources = context.resources
		val packageName = context.packageName
		if (indexCount > 0) {
			val imageNames = resources.getStringArray(
				typedArray.getResourceId(indexCount - 1, -1)
			)

			resourceIds = IntArray(imageNames.size)

			for (i in imageNames.indices) {
				val imageName = imageNames[i].substring(
					imageNames[i].lastIndexOf('/') + 1,
					imageNames[i].lastIndexOf('.')
				)

				resourceIds[i] = resources.getIdentifier(
					imageName,
					"drawable", packageName
				)
			}

			typedArray.recycle()
		}
	}

	private fun localStoragePermissionGranted(): Boolean =
		PermissionsPreference.permissionsGranted(
			mContext,
			mContext.resources.getStringArray(R.array.storage_permissions)
		)

	private fun googleServicesAvailable(): Boolean = GoogleApiAvailability.getInstance()
		.isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS

	private fun getFilteredCloudStorageEntries(): Array<CharSequence> {
		val localStorageFilteredEntries =
			if (localStoragePermissionGranted()) entries else entries.filter {
				it != mContext.getString(
					R.string.local_storage_string
				)
			}.toTypedArray()
		val googleDriveFilteredEntries =
			if (googleServicesAvailable()) localStorageFilteredEntries else localStorageFilteredEntries.filter {
				it != mContext.getString(
					R.string.google_drive_string
				)
			}.toTypedArray()
		return googleDriveFilteredEntries
	}

	private fun getFilteredCloudStorageEntryValues(): Array<CharSequence> {
		val localStorageFilteredEntries =
			if (localStoragePermissionGranted()) entryValues else entryValues.filter {
				it != mContext.getString(
					R.string.localStorageValue
				)
			}.toTypedArray()
		val googleDriveFilteredEntries =
			if (googleServicesAvailable()) localStorageFilteredEntries else localStorageFilteredEntries.filter {
				it != mContext.getString(
					R.string.googleDriveValue
				)
			}.toTypedArray()
		return googleDriveFilteredEntries
	}

	override fun onBindViewHolder(view: PreferenceViewHolder) {
		super.onBindViewHolder(view)
		val imageView = view.findViewById(R.id.iconImageView) as ImageView
		val iconResource = when (Preferences.getStringPreference(key, "")) {
			BeatPrompter.appResources.getString(R.string.googleDriveValue) -> R.drawable.ic_google_drive
			BeatPrompter.appResources.getString(R.string.dropboxValue) -> R.drawable.ic_dropbox
			BeatPrompter.appResources.getString(R.string.oneDriveValue) -> R.drawable.ic_onedrive
			BeatPrompter.appResources.getString(R.string.localStorageValue) -> R.drawable.ic_device
			BeatPrompter.appResources.getString(R.string.midi_usb_on_the_go_value) -> R.drawable.ic_usb
			BeatPrompter.appResources.getString(R.string.midi_native_value) -> R.drawable.midi
			BeatPrompter.appResources.getString(R.string.midi_bluetooth_value) -> R.drawable.ic_bluetooth
			BeatPrompter.appResources.getString(R.string.bluetoothModeBandLeaderValue) -> R.drawable.master0
			BeatPrompter.appResources.getString(R.string.bluetoothModeBandMemberValue) -> R.drawable.duncecap
			else -> R.drawable.blank_icon
		}
		imageView.setImageResource(iconResource)
	}

	fun forceUpdate() {
		if (!inForceUpdate) {
			try {
				inForceUpdate = true
				notifyChanged()
			} finally {
				inForceUpdate = false
			}
		}
	}

	companion object {
		const val GooglePlayServicesPackageName = "com.android.market"
	}
}
